### PR TITLE
Cut down on cold start times, support ESM in Lambda

### PR DESCRIPTION
**Pull request by raphael-theriault-swi**
Instead of always loading the instrumentation classes at the top of the file and only instantiating them if needed, we do both the loading and instantiating only when required, which significantly cuts down on cold start time (the runtime has a lot less parsing and evaluating to do) at the cost of some added code complexity. Given the massive improvements I think the compromise is more than worth it.
This also ends up making ESM faster than CJS so we might as well add support !

### CODE REVIEW

File: packages/instrumentations/src/index.ts [hunk 1]
```diff
@@ -18,136 +18,212 @@ import {
...
+import { type NetInstrumentation } from "@opentelemetry/instrumentation-net"
+import { type PgInstrumentation } from "@opentelemetry/instrumentation-pg"
+import { type PinoInstrumentation } from "@opentelemetry/instrumentation-pino"
+import { type RedisInstrumentation as RedisInstrumentationV2 } from "@opentelemetry/instrumentation-redis"
+import { type RedisInstrumentation as RedisInstrumentationV4 } from "@opentelemetry/instrumentation-redis-4"
+import { type RestifyInstrumentation } from "@opentelemetry/instrumentation-restify"
+import { type RouterInstrumentation } from "@opentelemetry/instrumentation-router"
+import { type SocketIoInstrumentation } from "@opentelemetry/instrumentation-socket.io"
+import { type TediousInstrumentation } from "@opentelemetry/instrumentation-tedious"
+import { type WinstonInstrumentation } from "@opentelemetry/instrumentation-winston"
```
Comment (cheempz) [inline packages/instrumentations/src/index.ts:57]: I think I get your epiphany--just import the type here (cheap and fast), these are used to build the list of instrumentation modules that at time of `getInstrumentations` will be checked whether enabled in config and actually be loaded? Very cool and it seems this optimization applies to non-lambda environment too!

Dumb noob question, it seems importing type is a "newish" typescript feature, does that translate into anything impacting the Node.js version that we support?

File: packages/instrumentations/src/index.ts [hunk 2]
```diff
@@ -18,136 +18,212 @@ import {
...
-    new (config?: InstrumentationConfig) => Instrumentation
-  >(INSTRUMENTATIONS)) {
-    let config = c[name]
+  const instrumentations = Object.entries(allConfigs)
+    .filter(([, config]: [unknown, InstrumentationConfig]) => {
+      // explicitly set "enabled" to false if that's the default
+      if (defaultDisabled) config.enabled ??= false
+      // filter out disabled instrumentations
+      return config.enabled !== false
+    })
```
Comment (cheempz) [inline packages/instrumentations/src/index.ts:192]: It seems we currently set `instrumentationsDefaultDisabled: boolean.default(IS_SERVERLESS)`, does this mean no instrumentation will be enabled in lambda?

I'm confused how the upstream https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/opentelemetry-instrumentation-aws-lambda instrumentation (and the aws-sdk instrumentation) are available in our test lambda functions if no instrumentation is enabled (maybe the existing layers had them all enabled?).  Also, the thing I mentioned about the upstream lambda instrumentation ignoring [.mjs handler filename](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts#L116-L128) -- not seeing how that's addressed in our custom distro, can you point me to how this works?

**Review by cheempz — APPROVED** [review]
Awesome improvement @raphael-theriault-swi! And really appreciate the Jira description and code comments that helped me understand changes.  I left several questions mainly to help more with my understanding :)

File: packages/instrumentations/src/index.ts (hunk 1 above)
Comment (raphael-theriault-swi) [reply packages/instrumentations/src/index.ts:57]: Type imports actually don't translate to anything in JavaScript, all of these statements are dropped when transpiling the code !

File: packages/instrumentations/src/index.ts (hunk 2 above)
Comment (raphael-theriault-swi) [reply packages/instrumentations/src/index.ts:192]: The AWS instrumentation configs are patched to take precedence over the defaults https://github.com/solarwinds/apm-js/blob/main/packages/sdk/src/patches/aws-lambda.ts

I'm actually not sure how the `.mjs` loading works given that code, I just know that it does work. Will look into what's going on there.

**Pull request merged**